### PR TITLE
*: directly report statistics to pd

### PR DESCRIPTION
**Pull request by nolouch**
refactor coprocessor and report statistics to pd worker

### CODE REVIEW

File: src/pd/metrics.rs [hunk 1]
```diff
@@ -20,4 +20,32 @@ lazy_static! {
             "Bucketed histogram of PD message send duration",
              exponential_buckets(0.0005, 10.0, 7).unwrap()
         ).unwrap();
+
+    pub static ref PD_REQ_COUNTER_VEC: CounterVec =
+        register_counter_vec!(
+            "tikv_raftstore_pd_request_sent_total",
```
Comment (overvenus) [inline src/pd/metrics.rs]: Now, it does not belong to `tikv_raftstore_` any more.

**Comment by siddontang** [comment]
PTAL @BusyJay

File: src/pd/metrics.rs (hunk 1 above)
Comment (siddontang) [reply src/pd/metrics.rs]: +1
We should also update the grafana too.

**Comment by siddontang** [comment]
Any update @nolouch

File: src/pd/pd.rs [hunk 2]
```diff
@@ -395,6 +426,21 @@ impl<T: PdClient> Runner<T> {
         handle.spawn(f);
         self.is_hb_receiver_scheduled = true;
     }
+
+    fn handle_coprocessor_read_stats(&mut self, read_stats: CopRequestStatistics) {
```
Comment (siddontang) [inline src/pd/pd.rs]: remove the name of coprocessor here, PD should not know it.

File: src/pd/pd.rs [hunk 3]
```diff
@@ -444,6 +498,7 @@ impl<T: PdClient> Runnable<Task> for Runner<T> {
             }
             Task::ReportSplit { left, right } => self.handle_report_split(handle, left, right),
             Task::ValidatePeer { region, peer } => self.handle_validate_peer(handle, region, peer),
+            Task::CopReadStats { read_stats } => self.handle_coprocessor_read_stats(read_stats),
```
Comment (siddontang) [inline src/pd/pd.rs]: ReadStats is fine.

File: src/coprocessor/endpoint.rs [hunk 4]
```diff
@@ -63,14 +64,14 @@ const OUTDATED_ERROR_MSG: &'static str = "request outdated.";
 
 const ENDPOINT_IS_BUSY: &'static str = "endpoint is busy";
 
-pub struct Host<R: CopSender + 'static> {
+pub struct Host {
```
Comment (siddontang) [inline src/coprocessor/endpoint.rs]: use another PR for refactoring coprocessor.

Comment (siddontang) [reply src/coprocessor/endpoint.rs]: Btw, you shouldn't need to let coprocessor know the PD worker directly. You can still pass a sender so you can test the coprocessor independently and easily.

File: src/coprocessor/endpoint.rs [hunk 5]
```diff
@@ -800,7 +792,8 @@ mod tests {
         let engine = engine::new_local_engine(TEMP_DIR, &[]).unwrap();
         let mut cfg = Config::default();
         cfg.end_point_concurrency = 1;
-        let mut end_point = Host::new(engine, worker.scheduler(), &cfg, MockCopSender::new());
+        let pd_worker = FutureWorker::new("test-pd-worker");
```
Comment (siddontang) [inline src/coprocessor/endpoint.rs]: how to check we have already sent the read statistics?

Comment (nolouch) [reply src/coprocessor/endpoint.rs]: maybe mock a pd  runner?

File: src/raftstore/store/metrics.rs [hunk 6]
```diff
@@ -129,20 +129,6 @@ lazy_static! {
              exponential_buckets(1.0, 2.0, 20).unwrap()
         ).unwrap();
 
-    pub static ref REGION_READ_KEYS_HISTOGRAM: Histogram =
```
Comment (siddontang) [inline src/raftstore/store/metrics.rs]: no this metric now?

File: src/pd/pd.rs [hunk 7]
```diff
@@ -34,6 +34,8 @@ use raftstore::store::Msg;
 use raftstore::store::util::{get_region_approximate_size, is_epoch_stale};
 use raftstore::store::store::StoreInfo;
 use raftstore::store::Callback;
+use coprocessor::CopRequestStatistics;
```
Comment (siddontang) [inline src/pd/pd.rs]: don't let PD know any coprocessor thing.

File: src/coprocessor/endpoint.rs (hunk 5 above)
Comment (siddontang) [reply src/coprocessor/endpoint.rs]: seem no need now.

File: src/pd/pd.rs [hunk 8]
```diff
@@ -241,6 +265,19 @@ impl<T: PdClient> Runner<T> {
...
+        stats.set_bytes_read(
+            self.store_stat.engine_total_bytes_read - self.store_stat.engine_last_total_bytes_read,
+        );
+        stats.set_keys_read(
+            self.store_stat.engine_total_keys_read - self.store_stat.engine_last_total_keys_read,
+        );
+        self.store_stat.engine_last_total_bytes_read = self.store_stat.engine_total_bytes_read;
+        self.store_stat.engine_last_total_keys_read = self.store_stat.engine_total_keys_read;
+
+        for peer in self.region_peers.values() {
```
Comment (siddontang) [inline src/pd/pd.rs]: define a local metric and flush after the for loop like https://github.com/pingcap/tikv/blob/master/src/raftstore/store/store.rs#L1692

Comment (siddontang) [reply src/pd/pd.rs]: I prefer handling write bytes/keys to PD worker too.

**Comment by siddontang** [comment]
Any update @nolouch

File: Cargo.toml [hunk 9]
```diff
@@ -80,7 +80,7 @@ git = "https://github.com/pingcap/kvproto.git"
 git = "https://github.com/pingcap/tipb.git"
 
 [dependencies.prometheus]
-version = "0.3"
+git = "https://github.com/pingcap/rust-prometheus.git"
```
Comment (siddontang) [inline Cargo.toml]: @overvenus 

Please update the crate.

File: src/raftstore/store/store.rs [hunk 10]
```diff
@@ -1693,34 +1676,13 @@ impl<T: Transport, C: PdClient> Store<T, C> {
         for peer in self.region_peers.values_mut() {
             peer.peer_stat.last_written_bytes = peer.peer_stat.written_bytes;
```
Comment (siddontang) [inline src/raftstore/store/store.rs]: can we save the last_written_bytes/keys in PD worker and update in the heartbeat msg?

**Comment by nolouch** [comment]
PTAL

File: src/pd/pd.rs [hunk 11]
```diff
@@ -389,25 +474,47 @@ impl<T: PdClient> Runnable<Task> for Runner<T> {
...
+                let (
+                    read_bytes_interval,
+                    read_keys_interval,
+                    written_bytes_interval,
+                    written_keys_interval,
+                ) = {
+                    let peer_stat = self.region_peers
+                        .entry(region.get_id())
+                        .or_insert_with(PeerStat::default);
+                    let read_bytes_interval = peer_stat.read_bytes - peer_stat.last_read_bytes;
```
Comment (siddontang) [inline src/pd/pd.rs]: delta may be better than interval

**Comment by siddontang** [comment]
Rest LGTM

PTAL @huachaohuang @zhangjinpeng1987

File: src/pd/pd.rs [hunk 12]
```diff
@@ -114,6 +157,8 @@ pub struct Runner<T: PdClient> {
     pd_client: Arc<T>,
     ch: SendCh<Msg>,
     db: Arc<DB>,
+    region_peers: HashMap<u64, PeerStat>,
```
Comment (zhangjinpeng1987) [inline src/pd/pd.rs]: When a peer is dropped in raftstore, here should keep updated.

**Pull request merged**